function_migrations.go

```diff
@@ -203,6 +204,137 @@ func migrateToSpecVersion(f Function, m migration) (Function, error) {
 	return f, nil
 }
 
+// This migration makes sure use the 1.0.0 format. To avoid unmarshalling issues with the old format
+// this migration needs to be executed first. Further migrations will operate on this new struct
+func migrateTo100Structure(f1 Function, m migration) (Function, error) {
+	// Load the Function using pertinent parts of the previous version's schema:
+	f0Filename := filepath.Join(f1.Root, FunctionFile)
+	bb, err := ioutil.ReadFile(f0Filename)
+	if err != nil {
+		return f1, err
+	}
+	f0 := migrateTo100_previousFunction{}
+	if err = yaml.Unmarshal(bb, &f0); err != nil {
+		return f1, err
+	}
+
+	f1.Name = f0.Name
+	f1.Runtime = f0.Runtime
+	f1.Template = f0.Template
+	f1.Registry = f0.Registry
+	f1.Image = f0.Image
+	f1.ImageDigest = f0.ImageDigest
+	f1.Invocation = f0.Invocation
+
```

Comment:
We shouldn't need to map every field, only those which have "moved":

1)  Assume `f1` (current version) is populated with everything that mapped over unchanged from the  previous version. (we're not in strict mode, so fields which no longer exist such as `.Git` should be ignored).
2) Load `func.yaml` _again_ (f0), but this time using the `previous version` schema (loads `.Git` etc)
3) map over anything which was moved (`.Git` -> `.Build.Git`).

This is because, at this granular resolution we should not have any unmarshalling errors _between_ migrations.  Those only occur when trying to jump f0 -> fN.  For example, if a field is removed then added back with a new schema (such as `.Build`, which used to be a string and is now a struct), that would cause an unmarshalling error during the initial unmarshalling, but when that fails, migrations are triggered and _between_ migrations there would be no error, because one migration removes the field, the next migration adds it back with the new type.

So I think you might want to rename `f1` back to `f` here like the other migrations; presume it is populated with everytyhing which has not changed (`Name`, `Runtime` etc), and then only use `f0` to grab the stuff that would be lost otherwise (all the fields relocated to the `XSpec` sub-structs).